flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/callback/SinkUserCallback.java

```diff
@@ -0,0 +1,36 @@
+package org.apache.flink.connector.pulsar.sink.callback;
+
+import javax.annotation.Nullable;
+import org.apache.flink.connector.pulsar.sink.writer.message.PulsarMessage;
+import org.apache.pulsar.client.api.MessageId;
+
+public interface SinkUserCallback<IN> extends AutoCloseable {
+    /**
+     * This method is called before the message is sent to the topic.
+     * The user can modify the message. By default, the same message will be returned.
+     * @param element the element received from the previous operator.
+     * @param message the message wrapper with the element already serialized.
+     * @param topic the pulsar topic or partition that the message will be routed to.
+     * @return the message to send to the pulsar topic, potentially a new message.
+     */
+    default PulsarMessage<?> beforeSend(IN element, PulsarMessage<?> message, String topic) {
```

Comment:
What's the scenario when users need to modify the message? If users modify the message then `element` will be inconsistent with `message`, do we need to call this out in the comment as caveat? The logic after beforeSend will still be using element.

Reply:
So this API takes inspiration from [`PulsarProducerInterceptor`](https://github.com/apache/pulsar/blob/master/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ProducerInterceptor.java).

A scenario where we would want to modify the `message` is: let's say you want to propagate some spans, or add extra tracing headers, you could add those in the message properties. For example if you want to do tracing downstream without the cost of serialization, the next step could simply get the info it needs from the message property bag that was added in the previous step. 

Good question about the potential conflict between `element` and `message`. The spirit of this was that we don't really modify the message, but "enrich" it for downstream consumers. Open to suggestions on how to address this.

I pass the `element` in the callback so the user can avoid the cost of extra serialization if what they need for enrichment is in the message body.

Reply:
Actually, giving it a bit more thought, it would be better to just return void in all the callback messages and not allow modifications. The ability to modify the `message` in the send callbacks is redundant since the user already controls the serializer, and they do can whatever they want there already.

Similarly on the source side, we can make the callbacks read-only. The user can make whatever transformations they want to the message in the deserializer, or a dedicated operator. There is no need for this to be done in a callback.